src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -0,0 +1,79 @@
+package viper.silver.plugin.standard.inline
+
+import viper.silver.ast.{Exp, Fold, LocalVar, Method, PredicateAccess, Program, Stmt, Unfold}
+
+trait InlineRewrite {
+
+  def inlinePredicates(method: Method, program: Program): Method = {
+    val expandedPres = method.pres.map { pre =>
+      expandPredicate(pre, program).fold(pre)(expandedPred => expandedPred)
+    }
+    val expandedPosts = method.posts.map { post =>
+      expandPredicate(post, program).fold(post)(expandedPred => expandedPred)
+    }
+    method.copy(name = method.name,
+      formalArgs = method.formalArgs,
+      formalReturns = method.formalReturns,
+      pres = expandedPres,
+      posts = expandedPosts,
+      body = method.body
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  def removeUnfoldFold(method: Method, program: Program): Method = {
+    val expandablePredicateIds = getExpandablePredicateIds(method.pres, program)
+    val rewrittenBody = method.body.map { body =>
+      val bodyWithRemovedUnfolds = removeUnfolds(body.ss, expandablePredicateIds)
+      val bodyWithRemovedUnfoldFolds = removeFolds(bodyWithRemovedUnfolds, expandablePredicateIds)
+      body.copy(ss = bodyWithRemovedUnfoldFolds, scopedDecls = body.scopedDecls)(
+        pos = body.pos, info = body.info, errT = body.errT
+      )
+    }
+    method.copy(name = method.name,
+      formalArgs = method.formalArgs,
+      formalReturns = method.formalReturns,
+      pres = method.pres,
+      posts = method.posts,
+      body = rewrittenBody
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  private[this] def expandPredicate(expr: Exp, program: Program): Option[Exp] =
+    expr match {
+      case pred: PredicateAccess => getPredicateBody(pred, program)
+      case _ => None
+    }
```

Comment:
Expressions might contain further expressions (e.g. a conjunction); we need to visit all subexpressions and expand all inner predicates. Something like this:

```scala
    ViperStrategy.CustomContext[Set[String]]({
      case (expr, args) => expr match {
        case pred: PredicateAccess =>
          (pred.predicateBody(program, args).getOrElse(pred), args)
        case quant: QuantifiedExp => ???
          // TODO: Continue traversing quant, but with context (args ++ quant.scopedDecls.map { _.name }.toSet)
      }
    }, method.scopedDecls.map { _.name }.toSet)
      // .recurseFunc(???) // TODO: Do we need to call this first?
      .execute(expr)
```